File: libs/shinkai-tools-builder/src/tool-definition.ts
import * as Ajv from 'ajv/dist/2019';

export interface ToolDefinition<
  TConfig extends Record<string, any>,
  TParameters extends Record<string, any>,
  TResult extends Record<string, any>,
> {
  id: string;
  name: string;
  description: string;
  author: string;
  keywords: string[];

  configurations: Ajv.JSONSchemaType<TConfig>;
  parameters: Ajv.JSONSchemaType<TParameters>;
  result: Ajv.JSONSchemaType<TResult>;
}

File: libs/shinkai-tools-runner/src/tools/tool_definition.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ToolDefinition {
    pub id: String,
    pub name: String,
    pub description: String,
    pub author: String,
    pub keywords: Vec<String>,
    pub configurations: Value,
    pub parameters: Value,
    pub result: Value,
    pub code: Option<String>,
}
